File: python/ttagy/parser.py
"""
TTAgy NDJSON 流式行事件解析器 (Python 实现与 Rust 100% 对齐)
"""
import json
from typing import List, Optional, Union, Dict, Any

class ParsedThinkingDelta:
    def __init__(self, delta: str):
        self.delta = delta
    def __repr__(self):
        return f"ParsedThinkingDelta({self.delta!r})"
    def __eq__(self, other):
        return isinstance(other, ParsedThinkingDelta) and self.delta == other.delta

class ParsedContentDelta:
    def __init__(self, delta: str):
        self.delta = delta
    def __repr__(self):
        return f"ParsedContentDelta({self.delta!r})"
    def __eq__(self, other):
        return isinstance(other, ParsedContentDelta) and self.delta == other.delta

class ParsedDone:
    def __init__(self, content: str, thinking_content: Optional[str] = None, prompt_tokens: Optional[int] = None, output_tokens: Optional[int] = None):
        self.content = content
        self.thinking_content = thinking_content
        self.prompt_tokens = prompt_tokens
        self.output_tokens = output_tokens
    def __repr__(self):
        return f"ParsedDone(content={self.content!r}, prompt_tokens={self.prompt_tokens}, output_tokens={self.output_tokens})"
    def __eq__(self, other):
        return (isinstance(other, ParsedDone) and self.content == other.content and 
                self.thinking_content == other.thinking_content and 
                self.prompt_tokens == other.prompt_tokens and 
                self.output_tokens == other.output_tokens)

class ParsedError:
    def __init__(self, code: str, message: str):
        self.code = code
        self.message = message
    def __repr__(self):
        return f"ParsedError(code={self.code!r}, message={self.message!r})"
    def __eq__(self, other):
        return isinstance(other, ParsedError) and self.code == other.code and self.message == other.message

ParsedItem = Union[ParsedThinkingDelta, ParsedContentDelta, ParsedDone, ParsedError]

class NdjsonParser:
    @staticmethod
    def parse_line_items(line: str) -> List[ParsedItem]:
        trimmed = line.strip()
        if not trimmed:
            return []
        try:
            val = json.loads(trimmed)
        except Exception:
            return []

        items: List[ParsedItem] = []
        ev_name = val.get("event") or val.get("type") or ""

        if ev_name == "step_update" and isinstance(val.get("step_update"), dict):
            step = val["step_update"]
            thought = step.get("thought_delta") or step.get("reasoning_delta") or step.get("thinking_delta")
            if thought and isinstance(thought, str):
                items.append(ParsedThinkingDelta(thought))
            text = step.get("text_delta") or step.get("content_delta")
            if text and isinstance(text, str):
                items.append(ParsedContentDelta(text))

        elif ev_name == "content":
            text = val.get("content") or val.get("text")
            if text and isinstance(text, str):
                items.append(ParsedContentDelta(text))

        elif ev_name == "message":
            msg = val.get("message")
            text = val.get("content") or val.get("text")
            if not text and isinstance(msg, dict):
                text = msg.get("content") or msg.get("text")
            if text and isinstance(text, str):
                items.append(ParsedContentDelta(text))

        elif ev_name in ("result", "done"):
            res_obj = val.get("result") if isinstance(val.get("result"), dict) else val
            status = res_obj.get("status", "")
            if status == "ERROR" or "error" in res_obj:
                err = res_obj.get("error")
                err_msg = err.get("message", "AGY worker error") if isinstance(err, dict) else str(err or "AGY worker error")
                items.append(ParsedError("AGY_ERROR", err_msg))
                return items

            content = (res_obj.get("response") or res_obj.get("content") or 
                       res_obj.get("text") or res_obj.get("structured_output") or "")
            if not isinstance(content, str):
                content = json.dumps(content) if content is not None else ""

            thinking = res_obj.get("thinking_content") or res_obj.get("thought")
            if thinking is not None and not isinstance(thinking, str):
                thinking = str(thinking)

            usage = res_obj.get("usage")
            prompt_tokens = None
            output_tokens = None
            if isinstance(usage, dict):
                prompt_tokens = usage.get("prompt_tokens")
                output_tokens = usage.get("completion_tokens") or usage.get("output_tokens")

            items.append(ParsedDone(
                content=content,
                thinking_content=thinking,
                prompt_tokens=prompt_tokens,
                output_tokens=output_tokens,
            ))

        elif ev_name == "error":
            err = val.get("error") or val.get("message") or "Unknown CLI error"
            err_msg = err.get("message", "Unknown error") if isinstance(err, dict) else str(err)
            items.append(ParsedError("CLI_ERROR", err_msg))

        return items

File: crates/ttagy-core/src/v1/parser.rs
//! NDJSON 流式行事件解析器 (Layer 2 协议解耦解析器)

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UsageMetadata {
    pub prompt_tokens: Option<usize>,
    pub output_tokens: Option<usize>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParsedStreamItem {
    ThinkingDelta(String),
    ContentDelta(String),
    Done {
        content: String,
        thinking_content: Option<String>,
        usage: Option<UsageMetadata>,
    },
    Error {
        code: String,
        message: String,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParsedChunk {
    ThinkingDelta(String),
    ContentDelta(String),
    Result(String),
    Error(String),
    Ignored,
}

pub struct NdjsonParser;

impl NdjsonParser {
    /// 完整流式解析：单行可能产出多个事件 (如同一更新中同时包含思考增量和正文增量)
    pub fn parse_line_items(line: &str) -> Vec<ParsedStreamItem> {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            return Vec::new();
        }

        let val: Value = match serde_json::from_str(trimmed) {
            Ok(v) => v,
            Err(_) => return Vec::new(),
        };

        let mut items = Vec::new();
        let ev_name = val.get("event")
            .or_else(|| val.get("type"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        match ev_name {
            "step_update" => {
                if let Some(step) = val.get("step_update") {
                    if let Some(thought) = step.get("thought_delta")
                        .or_else(|| step.get("reasoning_delta"))
                        .or_else(|| step.get("thinking_delta"))
                        .and_then(|v| v.as_str()) {
                        if !thought.is_empty() {
                            items.push(ParsedStreamItem::ThinkingDelta(thought.to_string()));
                        }
                    }
                    if let Some(delta) = step.get("text_delta")
                        .or_else(|| step.get("content_delta"))
                        .and_then(|v| v.as_str()) {
                        if !delta.is_empty() {
                            items.push(ParsedStreamItem::ContentDelta(delta.to_string()));
                        }
                    }
                }
            }
            "content" => {
                if let Some(chunk) = val.get("content").or_else(|| val.get("text")).and_then(|v| v.as_str()) {
                    if !chunk.is_empty() {
                        items.push(ParsedStreamItem::ContentDelta(chunk.to_string()));
                    }
                }
            }
            "message" => {
                let content = val.get("content")
                    .or_else(|| val.get("text"))
                    .or_else(|| val.get("message").and_then(|m| m.get("content").or_else(|| m.get("text"))))
                    .and_then(|v| v.as_str());
                if let Some(chunk) = content {
                    if !chunk.is_empty() {
                        items.push(ParsedStreamItem::ContentDelta(chunk.to_string()));
                    }
                }
            }
            "result" | "done" => {
                if let Some(res_obj) = val.get("result").or_else(|| Some(&val)) {
                    let status = res_obj.get("status").and_then(|v| v.as_str()).unwrap_or("");
                    if status == "ERROR" || res_obj.get("error").is_some() {
                        let err_msg = res_obj.get("error")
                            .and_then(|v| if v.is_string() { v.as_str() } else { v.get("message").and_then(|m| m.as_str()) })
                            .unwrap_or("AGY worker execution error");
                        items.push(ParsedStreamItem::Error {
                            code: "AGY_ERROR".to_string(),
                            message: err_msg.to_string(),
                        });
                        return items;
                    }

                    let content = res_obj.get("response")
                        .or_else(|| res_obj.get("content"))
                        .or_else(|| res_obj.get("text"))
                        .or_else(|| res_obj.get("structured_output"))
                        .and_then(|v| if v.is_string() { v.as_str() } else { None })
                        .unwrap_or("");

                    let thinking = res_obj.get("thinking_content")
                        .or_else(|| res_obj.get("thought"))
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());

                    let usage = res_obj.get("usage").map(|u| UsageMetadata {
                        prompt_tokens: u.get("prompt_tokens").and_then(|v| v.as_u64()).map(|n| n as usize),
                        output_tokens: u.get("completion_tokens").or_else(|| u.get("output_tokens")).and_then(|v| v.as_u64()).map(|n| n as usize),
                    });

                    items.push(ParsedStreamItem::Done {
                        content: content.to_string(),
                        thinking_content: thinking,
                        usage,
                    });
                }
            }
            "error" => {
                let err_msg = val.get("error")
                    .or_else(|| val.get("message"))
                    .and_then(|v| if v.is_string() { v.as_str() } else { v.get("message").and_then(|m| m.as_str()) })
                    .unwrap_or("Unknown CLI error");
                items.push(ParsedStreamItem::Error {
                    code: "CLI_ERROR".to_string(),
                    message: err_msg.to_string(),
                });
            }
            _ => {}
        }

        items
    }

    /// 兼容旧版单个 chunk 返回接口
    pub fn parse_line(line: &str) -> ParsedChunk {
        let items = Self::parse_line_items(line);
        for item in items {
            match item {
                ParsedStreamItem::ThinkingDelta(d) => return ParsedChunk::ThinkingDelta(d),
                ParsedStreamItem::ContentDelta(d) => return ParsedChunk::ContentDelta(d),
                ParsedStreamItem::Done { content, .. } => return ParsedChunk::Result(content),
                ParsedStreamItem::Error { message, .. } => return ParsedChunk::Error(message),
            }
        }
        ParsedChunk::Ignored
    }
}
